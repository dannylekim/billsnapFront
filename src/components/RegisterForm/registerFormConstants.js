const type_password = {type: "password"};

export const registerFormInputs = [
{
    placeholder: "Enter your first name",
    name: "firstName",
    type: "text"
},
{
    placeholder: "Enter your middle name", 
    name: "middleName",
    type: "text"
},
{
    placeholder: "Enter your last name",
    name: "lastName",
    type: "text"
},
{
    placeholder: "Enter your date of birth",
    name: "birthDate",
    type: "date",
},
{
    placeholder: "Enter your email",
    name: "email",
    type: "email",
},
{
    placeholder: "Enter your phone number",
    name: "phoneNumber",
    type: "tel",
},
{
    placeholder: "Enter a password",
    name: "password",
    ...type_password
},
{
    placeholder: "Verify password",
    name: "confirmPassword",
    ...type_password
},
{
    placeholder: "Enter your location",
    name: "location",
    type: "text"
},
];

/**
 * Missing
 */