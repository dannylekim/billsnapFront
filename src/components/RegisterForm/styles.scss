

div.register__form div.form-inputs{
    width:60%;
    margin-left: auto;
    margin-right: auto;
}

div.register__form .form-control{
    font-size:25px;
    margin-top: 25px;
    color: black;
    background-color: white;
}

span#input_error{
    color: red;
}