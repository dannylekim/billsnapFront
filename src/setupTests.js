// setup file
/** 
 * DON'T TOUCH THIS FILE.
 * IT'S REQUIRED TO TEST COMPONENTS
*/
import Enzyme, { shallow, render, mount } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer  from 'react-test-renderer';

configure({ adapter: new Adapter() });

const matches = children => expect(
    renderer.create(children).toJSON()
  ).toMatchSnapshot();
  
global.shallow = shallow;
global.render = render;
global.mount = mount;
global.matches = matches;
global.localStorage = {
  setItem: jest.fn()
};