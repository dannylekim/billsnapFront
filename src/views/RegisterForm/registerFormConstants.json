[
  {
    "placeholder": "First name",
    "name": "firstName",
    "type": "text",
    "autoComplete": "first-name"
  },
  {
    "placeholder": "Last name",
    "name": "lastName",
    "type": "text",
    "autoComplete": "last-name"
  },
  {
    "placeholder": "Email",
    "name": "email",
    "type": "email",
    "autoComplete": "username"
  },
  {
    "placeholder": "Password",
    "name": "password",
    "type": "password",
    "autoComplete": "new-password"
  },
  {
    "placeholder": "Confirm password",
    "name": "confirmPassword",
    "type": "password",
    "autoComplete": "new-password"
  }
]
