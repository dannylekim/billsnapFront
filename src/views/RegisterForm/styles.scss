div.form-inputs{
    // padding-top: 1%;
    width:70%;
    margin-left: auto;
    margin-right: auto;
}

div.register__form .form-control{
    font-size: 20px;
    margin-top: 35px;
    color: black;
    background-color: white; 
}

span#input_error{
    color: red;
}

h6 a.login-link{
    color:rgba(165, 199, 245, 0.801);
}