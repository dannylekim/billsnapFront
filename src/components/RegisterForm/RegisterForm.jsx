import React, { useState } from "react";
import PropType from "prop-types";
import {Tooltip } from "shards-react";
import {registerFormInputs} from "./registerFormConstants";
import LoginRegisterForm from "../forms/LoginRegisterForm";
import {register} from "../../utils/requests/UserRequests";
import "./styles.scss";

export const RegisterForm = ({handleButtonClick ,onChange,validInputs,user_credentials}) => {
    const formProps = {
        type: "register",
        form_className : "register__form",
        constants : registerFormInputs,
        handleButtonClick,
        onChange,
        validInputs,
        buttonValue: "Submit",
        user_credentials
    };
    return (
        <LoginRegisterForm formProps = {formProps}/>
    );
};

export default (props) => {
    //validInputs
    const [validFirstName, setValidFirstName] = useState(true);
    const [validMiddleName, setValidMiddleName] = useState(true);
    const [validLastName, setValidLastName] = useState(true);
    const [validDOB, setValidDOB] = useState(true); //no future dates
    const [validPhone, setValidPhone] = useState(true);  
    const [validEmail, setValidEmail] = useState(true);  
    const [validPasswordFormat, setValidPasswordFormat] = useState(true);
    const [validPassword, setValidPassword] = useState(true); //confirmed password is same
    const [validLocation, setValidLocation] = useState(true); 
    const [user_credentials, setUserCredential] = useState({}); 
   
    const validInputs = {
            "firstName"        : validFirstName, 
            "middleName"       : validMiddleName, 
            "lastName"         : validLastName,
            "birthDate"        : validDOB,
            "email"            : validEmail,
            "phoneNumber"      : validPhone,
            "password"         : validPasswordFormat,
            "confirm_password" : validPassword,
            "location"         : validLocation
    };

    const nameRegex = new RegExp(/^[_A-z]*((-|\s)*[_A-z])*$/);
    const phoneRegex = RegExp(/^[(]?[0-9]{3}[)]?[-\s]?[0-9]{3}[-\s']?[0-9]{4}$/);
    const emailRegex = new RegExp(/[\w-\.]+@([\w-]+\.)+[\w-]{2,4}/);
    const passwordRegex = new RegExp(/(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+*!=]).{8,20}/); 
    
    const validatePassword = (password, passwordToConfirm) =>  (password === passwordToConfirm);
    
    /**
     * 
     * Object that returns the proper valid input function. 
     * 
     * */
    const getValidationFunction = (name, value) => {
        let validationMap = new Map();
        (name === "firstName") && validationMap.set("firstName",setValidFirstName(nameRegex.test(value)));
        (name === "middleName") && validationMap.set("middleName",setValidMiddleName(nameRegex.test(value)));
        (name === "lastName") && validationMap.set("lastName",setValidLastName(nameRegex.test(value)));
        (name === "phoneNumber") && validationMap.set("phoneNumber",setValidPhone(phoneRegex.test(value)));
        (name === "email") && validationMap.set("email",setValidEmail(emailRegex.test(value)));
        (name === "password") && validationMap.set("password",setValidPasswordFormat(passwordRegex.test(value)));
        (name === "location") && validationMap.set("location",setValidLocation(nameRegex.test(value)));
        (name === "birthDate") && validationMap.set("birthDate",setValidDOB(new Date(value.split('-')[0], parseInt(value.split('-')[1]) -1 ,value.split('-')[2])  <= new Date()));// "yyyy-mm-dd"

        return validationMap.get([name]);
    };
    
    const onChange = event => {
        const {name, value} = event.target;
       
            if(name !== "confirm_password"){
                getValidationFunction(name,value);
                setUserCredential({...user_credentials, [name] : value});
            }else {
                const {password} = {...user_credentials};
                const confirm_password = value;
                setValidPassword( !validatePassword(password, confirm_password) ? false : true); //if password matchs
            };
    };

    const checkValidity = (boolean_value, checkState, name) =>  (checkState === boolean_value && user_credentials[[name]] && user_credentials[[name]] !== "");

    const handleButtonClick = (e) => {
        //if it exists and true
        e.preventDefault();

        (   checkValidity(true,validPassword, "password") && 
            checkValidity(true,validPasswordFormat, "password") && 
            checkValidity(true,validFirstName, "firstName") && 
            checkValidity(true,validLastName, "lastName") && 
            checkValidity(true,validEmail, "email") &&
            checkValidity(true,validPhone, "phoneNumber")
        ) ?        
        register(user_credentials).then(response => {
            alert(JSON.stringify(response))
        }) :  alert("form not validated")

        //Call the API... 
    };
  
    const conditions = [
        {
            condition : (checkValidity(false,validPassword, "password")),
            tool_tip_info : {
                open : !validPassword,
                id: "#confirm_password",
                error_message: "Password does not match"
            }
        },{
            condition : (checkValidity(false,validPasswordFormat, "password")),
            tool_tip_info : {
                open : !validPasswordFormat,
                id: "#password",
                error_message: "Invalid password (Must contain at least 1 upper case, lower case, number and symbol, and be between 8-20 characters."
            }
        },{
            condition : (checkValidity(false,validFirstName, "firstName")),
            tool_tip_info : {
                open : !validFirstName,
                id: "#firstName",
                error_message: "Invalid First Name, no numbers or special characters."
            }
        },{
            condition : (checkValidity(false,validMiddleName, "middleName")),
            tool_tip_info : {
                open : !validMiddleName,
                id: "#middleName",
                error_message: "Invalid Middle Name, no numbers or special characters."
            }
        },{
            condition : (checkValidity(false,validLastName, "lastName")),
            tool_tip_info : {
                open : !validLastName,
                id: "#lastName",
                error_message: "Invalid Last Name, no numbers or special characters."
            }
        },{
            condition : (checkValidity(false,validDOB, "birthDate")),
            tool_tip_info : {
                open : !validDOB,
                id: "#birthDate",
                error_message: "Can't be born in a future date."
            }
        },{
            condition : (checkValidity(false,validPhone, "phoneNumber")),
            tool_tip_info : {
                open : !validPhone,
                id: "#phoneNumber",
                error_message: "Must be 10 digit numbers (xxx) xxx xxxx (can be seperated by - or space)."
            }
        },{
            condition : (checkValidity(false,validEmail, "email")),
            tool_tip_info : {
                open : !validEmail,
                id: "#email",
                error_message: "Invalid Email Format."
            }
        }
    ];

    return (
        //Clean this up
    <div className="regForm">
      <RegisterForm handleButtonClick={handleButtonClick} onChange={onChange} validInputs={validInputs} user_credentials = {user_credentials}/>
        { conditions.map(condition => 
            condition.condition && 
            <Tooltip
                key={condition.tool_tip_info.id}
                open={condition.tool_tip_info.open}
                target={condition.tool_tip_info.id}>
                <span id="input_error">{condition.tool_tip_info.error_message}</span>
            </Tooltip> || null
        )}
    </div>
    );
};

RegisterForm.propTypes = {
    handleButtonClick: PropType.func.isRequired
  };