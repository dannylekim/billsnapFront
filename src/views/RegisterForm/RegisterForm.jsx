import React, { useState } from "react";
import PropType from "prop-types";
import {
  Tooltip,
  Alert,
  Button,
  Form,
  FormGroup,
  FormInput
} from "shards-react";
import * as data from "./registerFormConstants.json";
import { register } from "../../utils/requests/UserRequests";
import "./styles.scss";

export const RegisterForm = ({
  handleButtonClick,
  onChange,
  validInvalidByName,
  conditions
}) => {
  return (
    <Form>
      <div className="form-inputs">
        {data.registerFormInputs.map((inputs, key) => (
          <FormGroup key={key} onChange={onChange}>
            <FormInput
              invalid={validInvalidByName(inputs.name, "invalid")}
              valid={validInvalidByName(inputs.name, "valid")}
              className="mb-2"
              type={inputs.type}
              name={inputs.name}
              id={inputs.name}
              placeholder={inputs.placeholder}
              autoComplete={inputs.autoComplete}
            />
          </FormGroup>
        ))}
      </div>
      <FormGroup>
        <Button
          size="lg"
          pill
          theme="dark"
          onClick={event => handleButtonClick(event)}
          name="submit"
        >
          Submit
        </Button>
      </FormGroup>

      {conditions.map(
        condition => //document.getElement so that tests pass else DOM not found for id ... 
          ((condition.condition || condition.condition === false) && (document.getElementById(condition.name) && document.getElementById(condition.name)!== "")) && (
            <Tooltip
              placement="left"
              key={`#${condition.name}`}
              open={condition.toolTipInfo.open}
              target={`#${condition.name}`}
            >
              <span id="input_error">{condition.toolTipInfo.errorMessage}</span>
            </Tooltip>
          )
      )}
    </Form>
  );
};

export default props => {
  //validInputs
  const defaultError = {
    isOpen: false,
    alertType: "",
    alertMessage: ""
  };
  const [validFirstName, setValidFirstName] = useState(true);
  const [validLastName, setValidLastName] = useState(true);
  const [validEmail, setValidEmail] = useState(true);
  const [validPasswordFormat, setValidPasswordFormat] = useState(true);
  const [validPassword, setValidPassword] = useState(true); //confirmed password is same
  const [userCredentials, setUserCredential] = useState({
    firstName: "",
    lastName: "",
    email: "",
    password: "",
    confirmPassword: ""
  });
  const [alertNotification, setAlertNotification] = useState({
    ...defaultError
  });

  const validInputs = {
    firstName: validFirstName,
    lastName: validLastName,
    email: validEmail,
    password: validPasswordFormat,
    confirmPassword: validPassword
  };

  const nameRegex = new RegExp(/^[_A-z]*((-|\s)*[_A-z])*$/);
  const emailRegex = new RegExp(/[\w-.]+@([\w-]+).+[\w-]{2,4}/); //wrong!!!
  const passwordRegex = new RegExp(
    /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+*!=]).{8,20}/
  );

  /**
   * Boolean function that returns if password matches.
   * @param {String} password user's password
   * @param {String} passwordToConfirm confirmation password.
   */
  const validatePassword = (password, passwordToConfirm) =>
    password === passwordToConfirm;

  /**
   * This function aids the react-shard valid and invalid prop to display the errors.
   * Won't change the color of confirm password, since confirmPassword isn't part of user credential.
   * @param {String} name className
   * @param {String} type valid or invalid
   */
  const validInvalidByName = (name, type) => {
    return type === "invalid"
      ? !validInputs[[name]]
      : userCredentials[[name]] && userCredentials[[name]] !== ""
      ? validInputs[[name]]
      : false;
  };

  /**
   * @description Sets the alert or turns it off by calling the setter.
   * @param {String} triggerType - type of trigger success or error.
   */
  const triggerAlert = (triggerType, message) => {
    triggerType === "error"
      ? setAlertNotification({
          isOpen: true,
          alertType: "danger",
          alertMessage: message
        })
      : setAlertNotification({
          isOpen: true,
          alertType: "success",
          alertMessage: message
        });
  };
  /**
   * @description dismiss the alert. Sets the alert states back to default.
   */
  const dismissAlert = () => {
    setAlertNotification(defaultError);
  };
  /**
   * Returns the value of the key given from the map.
   * Map value is the result of the regex.
   * Object that returns the proper valid input function.
   * Lots of if statements.
   * @param {String} name - the input name.
   * @param {String} value - the value of the input field.
   * */
  const getValidationFunction = (name, value) => {
    switch (name) {
      case "firstName":
        return setValidFirstName(nameRegex.test(value));
      case "lastName":
        return setValidLastName(nameRegex.test(value));
      case "email":
        return setValidEmail(emailRegex.test(value));
      case "password":
        return setValidPasswordFormat(passwordRegex.test(value));
      case "confirmPassword":
        const { password } = { ...userCredentials };
        return setValidPassword(validatePassword(password, value));
      default:
        return null;
    }
  };
  /**
   * @description Handles on change events of the form. closes alert
   * @param {Event} event
   */
  const onChange = event => {
    const { name, value } = event.target;
    getValidationFunction(name, value);
    setUserCredential(prev => ({ ...prev, [name]: value }));
    setAlertNotification(defaultError);
  };
  /**
   * @description Check if condition is true or false. Shortcut to writting full condition.
   * document.getElementById is used when the form is already submitted but an error happened.
   * Since the confirmPassword is removed, it will no longer be in the userCredentials.
   * @param {Boolean} boolean_value check if true or false.
   * @param {Object} checkState the state to check against.
   * @param {String} name the name of the input field.
   */
  const checkValidity = (boolean_value, checkState, name) => {
    const inputField =
      name === "confirmPassword"
        ? document.getElementById("confirmPassword")
        : userCredentials[[name]];
    if (checkState === boolean_value && inputField && inputField !== "")
      return true;
    else {
      return false;
    }
  };
  /**
   * @description Handles form submittion and checks validity of inputs.
   * @param {Event} e The event after clicking on the button.
   */
  const handleButtonClick = async e => {
    e.preventDefault();
    if (
      checkValidity(true, validPassword, "confirmPassword") &&
      checkValidity(true, validPasswordFormat, "password") &&
      checkValidity(true, validFirstName, "firstName") &&
      checkValidity(true, validLastName, "lastName") &&
      checkValidity(true, validEmail, "email")
    ) {
      delete userCredentials["confirmPassword"]; //remove confirm password
      const response = await register(userCredentials);
      if (response.statusCode === 201)
        triggerAlert("success", "Registration Successful!");
      //TODO should redirect this.props.history.push("/dashboard");!!!
      //if user already exists
      else triggerAlert("error", response.message);
    } else {//blank inputs
      data.registerFormInputs.forEach(error => {
        if (userCredentials[error.name] === "")
          getValidationFunction(error.name, -1);
      });
      triggerAlert("error", "Form Not Validated"); //handle response
    }
  };
  //list of error messages / information for the Tool Tip component.
  const conditions = [
    {
      name:"confirmPassword",
      condition: checkValidity(false, validPassword, "confirmPassword"),
      toolTipInfo: {
        open: !validPassword,
        errorMessage: !userCredentials.confirmPassword
          ? "Cannot be blank!"
          : "Password does not match"
      }
    },
    {
      name:"password",
      condition: checkValidity(false, validPasswordFormat, "password"),
      toolTipInfo: {
        open: !validPasswordFormat,
        errorMessage: !userCredentials.password
          ? "Cannot be blank!"
          : "Must contain at least 1 upper case, lower case, number and symbol, and be between 8-20 characters."
      }
    },
    {
      name:"firstName",
      condition: checkValidity(false, validFirstName, "firstName"),
      toolTipInfo: {
        open: !validFirstName,
        errorMessage: !userCredentials.firstName
          ? "Cannot be blank!"
          : "No numbers or special characters."
      }
    },
    {
      name:"lastName",
      condition: checkValidity(false, validLastName, "lastName"),
      toolTipInfo: {
        open: !validLastName,
        errorMessage: !userCredentials.lastName
          ? "Cannot be blank!"
          : "No numbers or special characters."
      }
    },
    {
      name:"email",
      condition: checkValidity(false, validEmail, "email"),
      toolTipInfo: {
        open: !validEmail,
        errorMessage: !userCredentials.email
          ? "Cannot be blank!"
          : "Invalid Email Format."
      }
    }
  ];

  return (
    <div className="register__container">
      <Alert
        dismissible={dismissAlert}
        open={alertNotification.isOpen}
        className="mb-3"
        theme={alertNotification.alertType}
      >
        {alertNotification.alertMessage}
      </Alert>

      <RegisterForm
        handleButtonClick={handleButtonClick}
        onChange={onChange}
        validInputs={validInputs}
        userCredentials={userCredentials}
        validInvalidByName={validInvalidByName}
        conditions={[...conditions]}
      />
      <div>
        <h6>
          Have an account?
          <Button
            className="login-link"
            onClick={() => props.setFormType("login")}
          >
            Login to your account.
          </Button>
        </h6>
      </div>
    </div>
  );
};

RegisterForm.propTypes = {
  handleButtonClick: PropType.func.isRequired
};
