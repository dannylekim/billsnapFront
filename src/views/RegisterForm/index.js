import RegisterForm from "./RegisterForm";
import {connect} from "react-redux";
import {setUser} from "../../redux/actions/userActions";

const mapDispatchToProps = (dispatch) => ({
  setUser: (user = {}) => dispatch(setUser(user)),
});

export default connect(null, mapDispatchToProps)(RegisterForm);
