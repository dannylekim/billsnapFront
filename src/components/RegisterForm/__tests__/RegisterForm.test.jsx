import React from 'react';
import { RegisterForm } from '../RegisterForm.jsx';
import { shallow } from "enzyme";
import {matches} from "../../../setupTests";

describe('RegisterForm', () => {
    let wrapper;
    let handleMockFunction;
  
    beforeEach(() => {
      handleMockFunction = jest.fn();
      wrapper = shallow(
        <RegisterForm handleButtonClick={handleMockFunction} validInvalidByName={handleMockFunction}/>
      );
    });
  
    afterEach(() => {
      handleMockFunction.mockRestore();
    });
  
    describe("render", () => {
      describe("snapshots ðŸ“¸", () => {
        it("Register should match snap shot", () => {
          matches(<RegisterForm handleButtonClick={handleMockFunction} validInvalidByName={handleMockFunction}/>);
        });
      });
  
      describe("components", () => {});
    });
  
    describe("functions", () => {});
});