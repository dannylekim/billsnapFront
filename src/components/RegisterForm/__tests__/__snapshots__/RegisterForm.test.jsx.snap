// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm render snapshots ðŸ“¸ Register should match snap shot 1`] = `
<div
  className="register__form"
>
  <form
    className=""
  >
    <div
      className="form-inputs"
    >
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="firstName"
          name="firstName"
          placeholder="Enter your first name"
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="middleName"
          name="middleName"
          placeholder="Enter your middle name"
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="lastName"
          name="lastName"
          placeholder="Enter your last name"
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="birthDate"
          name="birthDate"
          placeholder="Enter your date of birth"
          type="date"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="email"
          name="email"
          placeholder="Enter your email"
          type="email"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="phoneNumber"
          name="phoneNumber"
          placeholder="Enter your phone number"
          type="tel"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="password"
          name="password"
          placeholder="Enter a password"
          type="password"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="confirmPassword"
          name="confirmPassword"
          placeholder="Verify password"
          type="password"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="location"
          name="location"
          placeholder="Enter your location"
          type="text"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-dark btn-lg btn-pill"
        name="submit"
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
